### PR TITLE
Simplify `cupy.fuse` tests

### PR DESCRIPTION
**Pull request by asi1024**
code fix for `tests/cupy_tests/core_tests/test_fusion.py`

Depends on #2396.

### CODE REVIEW

File: tests/cupy_tests/core_tests/fusion_tests/test_array.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+
+
+class FusionArrayTestBase(unittest.TestCase):
+
+    def _get_argument(self, xp, dtype, seed, value_type):
+        dtype = numpy.dtype(dtype)
+        if value_type == 'array':
+            x = testing.shaped_random((3, 4), xp, dtype, scale=5, seed=seed)
+            # Avoid zero-division
+            x[x == 0] = 1
```
Comment (niboshi) [inline tests/cupy_tests/core_tests/fusion_tests/test_array.py]: How about doing this only when necessary (in division tests), or adding TODO for that?

File: tests/cupy_tests/core_tests/fusion_tests/test_array.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+    def _is_signed_int(self, x):
+        return isinstance(x, int) or x.dtype.kind == 'i'
+
+    @testing.for_int_dtypes_combination(
+        names=('dtype1', 'dtype2'), no_bool=True)
+    @fusion_utils.check_fusion()
+    def test_operator(self, xp, dtype1, dtype2):
+        def func(x, y):
+            if ((self._is_uint64(x) and self._is_signed_int(y))
+                    or self._is_uint64(y) and self._is_signed_int(x)):
```
Comment (niboshi) [inline tests/cupy_tests/core_tests/fusion_tests/test_array.py]: Add parentheses around the `and` expression for readability.

File: tests/cupy_tests/core_tests/fusion_tests/test_function.py [hunk 3]
```diff
@@ -0,0 +1,553 @@
...
+            return a - b * g(x - 2, x - 3, -y)
+
+        x = testing.shaped_arange((3, 3), xp, dtype)
+        y = testing.shaped_arange((3, 3), xp, dtype)
+        return h(x, y)
+
+
+class TestFusionCompile(unittest.TestCase):
+
+    # TODO(asi1024): Support it
```
Comment (niboshi) [inline tests/cupy_tests/core_tests/fusion_tests/test_function.py]: Be more specific about what to support.
What's `_compile_from_dtypes`?

Comment (niboshi) [reply tests/cupy_tests/core_tests/fusion_tests/test_function.py]: Oh this is not directly related to this PR.

Comment (asi1024) [reply tests/cupy_tests/core_tests/fusion_tests/test_function.py]: Removed this commented test, because it was a test for a private method.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Please check test failures.

**Comment by asi1024** [comment]
@niboshi Fixed. PTAL.

**Comment by niboshi** [comment]
Thanks
Jenkins, test this please

**Comment by niboshi** [comment]
Hmm, there are still errors.

**Comment by asi1024** [comment]
pfnCI, test this please.

**Comment by niboshi** [comment]
Is it ready for review?

**Comment by asi1024** [comment]
@niboshi Yes. Could you take a look?

File: tests/cupy_tests/core_tests/fusion_tests/fusion_utils.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+def check_fusion(
+        generate_inputs_name='generate_inputs',
+        generate_inputs_args=None,
+        check_array=None,
+        check_array_kwargs=None,
+        accept_error=()):
+    """Decorator for tests for ``cupy.fuse``.
+
+    This decorator checks
```
Comment (niboshi) [inline tests/cupy_tests/core_tests/fusion_tests/fusion_utils.py]: ?

File: tests/cupy_tests/core_tests/fusion_tests/fusion_utils.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            ret = None
+            err = e
+        return ret, err
+
+    def check_result(xp, actual, expected):
+        ret_a, err_a = actual
+        ret_e, err_e = expected
+
+        if err_e is None:
+            # Success
```
Comment (niboshi) [inline tests/cupy_tests/core_tests/fusion_tests/fusion_utils.py]: This comment is misleading. "Exception not raised" is better.

File: tests/cupy_tests/core_tests/fusion_tests/fusion_utils.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+            assert isinstance(actual, dict)
+            for key, expected_value in expected:
+                actual_value = actual.pop(key)
+                check_array(xp, actual_value, expected_value)
+            assert len(actual) == 0
+
+        else:
+            assert False
+
+    def run(func, args, kwargs):
```
Comment (niboshi) [inline tests/cupy_tests/core_tests/fusion_tests/fusion_utils.py]: What is this function for? Please name it more explicitly and/or perhaps write a comment.

File: tests/cupy_tests/core_tests/fusion_tests/fusion_utils.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+    def deco(func):
+        def wrapper(self, **generate_inputs_kwargs):
+            generate_inputs = getattr(self, generate_inputs_name)
+
+            impl_np = func(self, numpy, **generate_inputs_kwargs)
+            impl_cp = func(self, cupy, **generate_inputs_kwargs)
+
+            # TODO(imanishi): Fix these workaround after `cupy.fuse`
+            # supports lambda function.
```
Comment (niboshi) [inline tests/cupy_tests/core_tests/fusion_tests/fusion_utils.py]: It's not clear what are "these workaround".

File: tests/cupy_tests/core_tests/fusion_tests/fusion_utils.py (hunk 4 above)
Comment (asi1024) [reply tests/cupy_tests/core_tests/fusion_tests/fusion_utils.py]: Sorry, I have forgotten to complete this comment.

File: tests/cupy_tests/core_tests/fusion_tests/fusion_utils.py (hunk 5 above)
Comment (asi1024) [reply tests/cupy_tests/core_tests/fusion_tests/fusion_utils.py]: Fixed as you commented.

File: tests/cupy_tests/core_tests/fusion_tests/fusion_utils.py (hunk 6 above)
Comment (asi1024) [reply tests/cupy_tests/core_tests/fusion_tests/fusion_utils.py]: Renamed to `call` and wrote comments.

File: tests/cupy_tests/core_tests/fusion_tests/fusion_utils.py (hunk 7 above)
Comment (asi1024) [reply tests/cupy_tests/core_tests/fusion_tests/fusion_utils.py]: Enhanced the comment.

**Comment by asi1024** [comment]
@niboshi PTAL.

**Comment by niboshi** [comment]
Thanks.
Jenkins, test this please

**Comment by niboshi** [comment]
There are two test failures:
```
15:45:37 FAIL tests/cupy_tests/fft_tests/test_fft.py::TestFft_param_6::test_ifft
```
```
15:36:41 FAIL tests/cupy_tests/core_tests/fusion_tests/test_function.py::TestFusionOutArg::test_outarg_add
```

**Comment by niboshi** [comment]
The FFT error is unrelated to this PR: #2441

File: tests/cupy_tests/core_tests/fusion_tests/test_function.py [hunk 8]
```diff
@@ -0,0 +1,540 @@
...
+    def generate_inputs(self, xp, dtype1, dtype2):
+        x = testing.shaped_random((3, 4), xp, dtype1, scale=10, seed=0)
+        y = testing.shaped_random((3, 4), xp, dtype1, scale=10, seed=1)
+        z = testing.shaped_random((3, 4), xp, dtype2, scale=10, seed=2)
+        return (x, y, z), {}
+
+    @testing.for_all_dtypes_combination(
+        names=('dtype1', 'dtype2'), full=True, no_complex=True)
+    @fusion_utils.check_fusion(accept_error=TypeError)
+    @testing.with_requires('numpy>=1.13')
```
Comment (asi1024) [inline tests/cupy_tests/core_tests/fusion_tests/test_function.py]: Note: Casting rule have changed from NumPy 1.13 ([Release notes](https://docs.scipy.org/doc/numpy-1.13.0/release.html#default-casting-rule-change))

**Comment by asi1024** [comment]
@niboshi PTAL.

File: tests/cupy_tests/core_tests/fusion_tests/test_function.py [hunk 9]
```diff
@@ -141,7 +141,7 @@ def func(x):
     @fusion_utils.check_fusion(generate_inputs_name='dtype_combination')
     def test_reduction_pairwise_premap(self, xp, dtype1, dtype2):
         def func(x, y):
-            return xp.sum(xp.sqrt(x - y))
```
Comment (niboshi) [inline tests/cupy_tests/core_tests/fusion_tests/test_function.py]: Is this test newly introduced in this PR?

File: tests/cupy_tests/core_tests/fusion_tests/test_function.py (hunk 8 above)
Comment (niboshi) [reply tests/cupy_tests/core_tests/fusion_tests/test_function.py]: Why hasn't this been a problem before this PR?

File: tests/cupy_tests/core_tests/fusion_tests/test_function.py (hunk 9 above)
Comment (asi1024) [reply tests/cupy_tests/core_tests/fusion_tests/test_function.py]: Yes.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
CI failure is unrelated.

**Pull request merged**